gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -537,8 +539,32 @@ public void createDirectory(Path dir, FileAttribute<?>... attrs) {
    */
   @Override
   public DirectoryStream<Path> newDirectoryStream(Path dir, Filter<? super Path> filter) {
-    // TODO: Implement me.
-    throw new UnsupportedOperationException();
+    CloudStoragePath cloudPath = checkPath(dir);
+    checkNotNull(filter);
+    String prefix = cloudPath.toString();
+    Iterator<Blob> iterator = storage.list(cloudPath.bucket(), Storage.BlobListOption.prefix(prefix)).iterateAll();
+    final List<Path> ret = new ArrayList<>();
+    while (iterator.hasNext()) {
+      Path p = (Path)(cloudPath.getFileSystem().getPath(iterator.next().name()));
+      try {
+        if (filter.accept(p)) {
+          ret.add(p);
+        }
+      } catch (IOException boom) {
+        // The doc says we should throw this in the iterator, but since we're prefilling
+        // the list this is the next best thing.
+        throw new DirectoryIteratorException(boom);
+      }
+    }
+    return new DirectoryStream<Path>() {
+      @Override
+      public Iterator<Path> iterator() {
+        return ret.iterator();
+      }
+      @Override
+      public void close() throws IOException {
```

Comment:
human-behind- @mziccard : do you really want me to put a comment saying "// close does nothing because there's nothing to close." atop the empty close() method? I'm guessing no, and this issue is a false positive.